### ISSUE TITLE
no-useless-escape: Special case for RexExp literal character sets.

### ISSUE THREAD

**Issue by JHawkley** [eslint/eslint#12856]
**What rule do you want to change?**
`no-useless-escape`

Likely related to pull-request #7425.

**Does this change cause the rule to produce more or fewer warnings?**
Fewer warnings.

**How will the change be implemented? (New option, new default behavior, etc.)?**
There's a special case for regular expression literals that might be worth excepting from this rule: escaping an opening-square-bracket character `[` that is used within a character set.

This ends up being a bit confusing due to asymmetry.  A `[` does not need to be escaped but a `]` must be escaped.

I suspect most developers would escape in this situation, by default, so making it the default behavior might be best and should not make any existing code that was lint-free linty.

An option could be provided to restore absolute strictness, for the machines among us.

**Please provide some example code that this change will affect:**
I came across the following in an older code-base that is integrating ESLint into its build process.  It's a character set that happens to be matching square-brackets.

```js
name = name.replace(/[\[\]]/g, "\\$&");
```

**What does the rule currently do for this code?**
It flags `\[` as being a useless escape.  While this is true (it still works fine when left unescaped), what ESLint would prefer, `/[[\]]/g`, is more difficult for a human to parse at first glance (not to say the original is cakewalk).

**What will the rule do after it's changed?**
It would quietly permit the escape of opening-square-brackets within character sets in regular expression literals.  The escape is useless from a syntactical perspective, but it is not useless from a legibility perspective.

**Are you willing to submit a pull request to implement this change?**
I have not yet investigated how difficult the change would be to make and the time investment required.  I can only say, "perhaps".

**Comment by mdjermanovic** [comment]
Hi @JHawkley, thanks for the enhancement proposal!

I think this is a reasonable request, but new default behavior as proposed looks to me like a breaking change since this isn't a bug.

I'd vote for an option in this rule: an array of allowed characters.

**Comment by kaicataldo** [comment]
Thanks for the issue! I’m personally not convinced we should add this to the core rule as this seems like unexpected behavior and a non-trivial increase in the complexity of the API. This is one of the few core rules we have that doesn’t have any options, and I think that makes sense given the rule’s goal. If it doesn’t need to be escaped, I think the rule should warn.

Is this style supported by any popular style guides?

**Comment by JHawkley** [comment]
I'm not familiar enough with Regex style guides to say.  I'm mostly guided by common sense, and my common sense tells me that if `]` needs an escape, then it is sensible to apply the same to `[`.  These two characters have a strong relation in developer's minds.

I argue that the strictest, most correct interpretation of a rule is not always the best interpretation for humans.

A regular-expression literal is syntax, unlike characters in strings and template literals.  Granting a bit of flexibility when it comes to the styling of syntax is something ESLint has been doing for some time.  In this case, if allowing the escape of `[` in a character set makes the code easier to interpret, it seems like a good exception to allow.

I haven't seen any reason to make any other exceptions, though.  I tried a few other expressions that I wondered might create a similar issue, but I couldn't find any.  So I'd actually recommend against generalizing this enhancement.  Far as I can tell, the rule pushes toward a more intuitive expression every other time.

And making this enhancement an option, rather than the default, is fine by me.  You do make it sound like adding option support is not a trivial change, though.  I'd still prefer to not have to add an `eslint-disable` to every regular-expression literal that runs afoul of this, though.  I do still want to lint for any **other** unnecessary escapes.

**Comment by kaicataldo** [comment]
I'm happy to defer to the consensus of the rest of the team on this.

I personally think the overhead of knowing the arbitrary exceptions a rule has is more overhead for developers, so do not think this should be the default behavior.  If we do implement this, I agree that a generic option that is an array of exceptions makes sense.
